clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java

```diff
@@ -121,8 +122,9 @@ public LeaderAndIsrRequest(Struct struct, short version) {
             List<Integer> replicas = new ArrayList<>(replicasArray.length);
             for (Object r : replicasArray)
                 replicas.add((Integer) r);
+            boolean isNew = partitionStateData.hasField(IS_NEW_KEY_NAME) ? partitionStateData.getBoolean(IS_NEW_KEY_NAME) : false;
```

Comment:
Note that this only matters when there is disk failure. Let me try to explain it based in the following different scenarios:

1) All brokers are running old code. The choice of the default value doesn't matter.

2) Brokers are rolling upgraded to use the new code but the IBP is still old.

Here is the concern with setting default value to true. Suppose the broker is running new code and partition p1 is in an offline log directories of this broker. If controller is running old code and sends LeaderAndIsrRequest asking this broker to be leader of p1, the broker will re-create partition p1 on a good log directory. The follower will truncate the log for p1 and the data will be lost. And KIP-112 currently doesn't handle it the case that the offline disk is repaired and used again with p1 still in it.

On the other hand, with default value set to false, broker will refuses to become leader for p1, controller will not handle re-elect leader for p1 because it is running old code. During this period partition will become unavailable because the producer/consumer will send request to this broker and receive NotLeadderForPartitionException. However, this will only happen for a short period of time and the problem will be solved once the controller is moved to a broker which is running new code.

Thus in this case, it is safer to set default value to false.

3) All brokers are running new code but some brokers are still using old IBP.

Suppose controller sends LeaderAndIsrRequest for a new partition p1 to a broker with offline log directory, the broker will refuse to create log for p1 and return error in LeaderAndIsrResponse. Since the controller is running new code, it will handle the error properly by re-electing leader for this partition. In the rare scenario that all brokers elected to be replicas for this partition have offline disk, the partition will be offline. But the chance of this happening is low and this will be addresses once the IBP is fully upgraded.

4) All brokers are running new code with the new IBP.

In this case the LeaderAndIsrRequest will explicitly specify the isNew field and the default value doesn't matter.

Does this make sense?